File: app/src/main/java/com/omar/retromp3recorder/app/recorder/FilePathGenerator.java
package com.omar.retromp3recorder.app.recorder;

import android.content.Context;

import com.omar.retromp3recorder.app.utils.NotUnitTestable;

import javax.inject.Inject;

@NotUnitTestable
public class FilePathGenerator {
    private static final String MP3_EXTENSION = ".mp3";
    private static final String VOICE_RECORD = "voice_record";

    private final Context context;

    @Inject
    public FilePathGenerator(Context context) {
        this.context = context;
    }

    public String generateFilePath() {
        String fileName = VOICE_RECORD + MP3_EXTENSION;
        return context.getExternalCacheDir() + "/" + fileName;
    }
}

File: utils/src/main/java/com/omar/retromp3recorder/utils/FilePathGenerator.kt
package com.omar.retromp3recorder.utils

import android.content.Context
import android.os.Environment
import java.io.File
import javax.inject.Inject

interface FilePathGenerator {
    fun generateFilePath(): String
    val fileDirs: List<String>
}

@Suppress("DEPRECATION")
class FilePathGeneratorImpl @Inject constructor(
    private val context: Context
) : FilePathGenerator {
    override fun generateFilePath(): String {
        return fileDirs.first { File(it).exists() }
    }

    override val fileDirs: List<String>
        get() = listOf(
            "${Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC)}/RetroMp3Recorder",
            context.externalCacheDir.toString(),
            context.cacheDir.toString()
        )
}
